### PR TITLE
RDR to setup backup and restore using acm

### PR DESCRIPTION
**Pull request by prsurve**
This PR Handles Setting up acm backup and recovery feature

### CODE REVIEW

File: ocs_ci/deployment/deployment.py [hunk 1]
```diff
@@ -2816,6 +2822,31 @@ def deploy(self):
...
+        # Reconfigure OADP on all ACM clusters
+        for i in acm_indexes:
+            config.switch_ctx(i)
+            create_dpa(self.meta_obj.bucket_name)
+        # Only on the active hub enable managedserviceaccount-preview
+        config.switch_acm_ctx()
+        enable_managed_serviceaccount()
+
 
 class MDRMultiClusterDROperatorsDeploy(MultiClusterDROperatorsDeploy):
```
Comment (am-agrawa) [inline ocs_ci/deployment/deployment.py:2851]: Can we keep this class name generic as DR?

File: ocs_ci/helpers/dr_helpers.py [hunk 2]
```diff
@@ -1082,3 +1095,213 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+        config.switch_ctx(index)
+        try:
+            run_cmd(f"oc create namespace {constants.ACM_HUB_BACKUP_NAMESPACE}")
+        except CommandFailed as ex:
+            if "already exists" in str(ex):
+                logger.warning("Namespace already exists!")
+            else:
+                raise
+        try:
+            run_cmd(cmd)
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:1208]: Pls add a validation for the same

```
amagrawa:~$ oc get secrets -n open-cluster-management-backup | grep cloud-credentials
cloud-credentials                                  Opaque                                1      22m

```

File: ocs_ci/helpers/dr_helpers.py [hunk 3]
```diff
@@ -1082,3 +1095,213 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    run_cmd(f"oc create -f {oadp_yaml.name}")
+    # Validation
+    validate_dpa()
+
+
+@retry((CommandFailed, ACMClusterConfigurationException), tries=10, delay=10)
+def validate_dpa():
+    """
+    Validate
+    1. 3 restic pods
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:1266]: Same, restic pods are no longer available I guess. Pls check and confirm

File: ocs_ci/helpers/dr_helpers.py [hunk 4]
```diff
@@ -1082,3 +1095,213 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    # Check backupstoragelocation resource in "Available" phase
+    backupstorage = ocp.OCP(
+        kind="BackupStorageLocation",
+        resource_name="default",
+        namespace=constants.ACM_HUB_BACKUP_NAMESPACE,
+    )
+    resource = backupstorage.get()
+    if resource["status"].get("phase") != "Available":
+        raise ACMClusterConfigurationException(
+            "Backupstoragelocation resource is no in 'Avaialble' phase"
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:1305]:
```suggestion
            "Backupstoragelocation resource is not in 'Available' phase"
```

File: ocs_ci/ocs/constants.py [hunk 5]
```diff
@@ -921,13 +921,11 @@
...
-MDR_VELERO_POD_COUNT = 1
+RESTIC_POD_COUNT = 3
+VELERO_POD_COUNT = 1
 MDR_DPA = "dpa-1"
-MDR_MULTICLUSTER_ENGINE = "multiclusterengine"
-MDR_BACKUP_SCHEDULE_YAML = os.path.join(
-    TEMPLATE_MULTICLUSTER_DIR, "backupschedule.yaml"
-)
+MULTICLUSTER_ENGINE = "multiclusterengine"
+BACKUP_SCHEDULE_YAML = os.path.join(TEMPLATE_MULTICLUSTER_DIR, "backupschedule.yaml")
```
Comment (am-agrawa) [inline ocs_ci/ocs/constants.py:928]: backupschedule.yaml needs changes, raised- https://github.com/red-hat-storage/ocs-ci/issues/9554

File: ocs_ci/helpers/dr_helpers.py [hunk 6]
```diff
@@ -1082,3 +1095,213 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+        )
+    logger.info("Dataprotection application successful")
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:1307]: Post this, we should also add a check for policy compliance on both ACMs and proceed with hub recovery if it's complaint.

```
amagrawa:~$ oc get policy backup-restore-enabled -n open-cluster-management-backup
NAME                     REMEDIATION ACTION   COMPLIANCE STATE   AGE
backup-restore-enabled   inform               Compliant          10m
```

File: ocs_ci/ocs/constants.py (hunk 5 above)
Comment (prsurve) [reply ocs_ci/ocs/constants.py:928]: This will be taken in different PR

File: ocs_ci/deployment/deployment.py (hunk 1 above)
Comment (prsurve) [reply ocs_ci/deployment/deployment.py:2851]: This class belongs to MDR So we cannot rename it

File: ocs_ci/ocs/constants.py [hunk 7]
```diff
@@ -921,13 +921,11 @@
 RDR_MODE = "regional-dr"
 MDR_MODE = "metro-dr"
 MDR_DR_POLICY = "odr-policy-mdr"
-MDR_RESTIC_POD_COUNT = 3
-MDR_VELERO_POD_COUNT = 1
+RESTIC_POD_COUNT = 3
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/ocs/constants.py:934]: Yes it does not it's be replaced with node-agent

File: ocs_ci/helpers/dr_helpers.py [hunk 8]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    compliance_output = ocp.OCP(
+        kind=constants.ACM_POLICY,
+        resource_name=resource_name,
+        namespace=resource_namespace,
+    )
+    compliance_status = compliance_output.get()
+    if compliance_status["status"]["compliant"] == compliance_state:
+        logger.info("Compliance status Matches ")
+    else:
+        raise ResourceWrongStatusException("Compliance status does not matches")
```
Comment (am-agrawa) [inline ocs_ci/helpers/dr_helpers.py:1355]:
```suggestion
        raise ResourceWrongStatusException("Compliance status does not match")
```

**Review by am-agrawa — APPROVED** [review]
Just a small typo, and the description of the PR is missing. Rest LGTM.

**Review by am-agrawa — APPROVED** [review]
LGTM

**Comment by am-agrawa** [comment]
Can we also add the label on passive hub post hub recovery to enable VolumeSync.DelayAlerts?

`oc label namespace openshift-operators openshift.io/cluster-monitoring='true'`

File: ocs_ci/deployment/deployment.py [hunk 9]
```diff
@@ -2933,177 +2979,40 @@ def deploy(self):
...
-            )
-        logger.info("Dataprotection application successful")
+            enable_managed_serviceaccount()
 
-    def create_generic_credentials(self):
-        s3_cred_str = (
-            "[default]\n"
-            f"aws_access_key_id={self.meta_obj.access_key}\n"
-            f"aws_secret_access_key={self.meta_obj.secret_key}\n"
+        validate_policy_compliance_status(
```
Comment (am-agrawa) [inline ocs_ci/deployment/deployment.py:3011]: The policy would be complaint on active when all the backups are created and on passive when all the backups are created matching that of active. So this check should be done post creating `backschedule.yaml` resource on the active hub with proper timeouts and re-tries.

Comment (prsurve) [reply ocs_ci/deployment/deployment.py:3011]: i will remove this verification step from here

**Review by am-agrawa — DISMISSED** [review]

**Comment by am-agrawa** [comment]
@parikshithb @Akarsha-rai @Shrivaibavi @hnallurv @avd-sagare Kindly spare some time and review the PR. Thanks!

File: ocs_ci/helpers/dr_helpers.py [hunk 10]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
         raise UnexpectedBehaviour(
             f"DRPolicy is not in succeeded or validated state: {status}"
         )
+
+
+def enable_cluster_backup():
```
[the remark this answers is not in the document]
Comment (prsurve) [reply ocs_ci/helpers/dr_helpers.py:1103]: it's because code is shared with MDR and RDR

File: ocs_ci/helpers/dr_helpers.py [hunk 11]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    )
+    yaml_serialized = yaml.dump(resource_dict)
+    mch_resource_yaml.write(yaml_serialized)
+    mch_resource_yaml.flush()
+    run_cmd(f"oc apply -f {mch_resource_yaml.name}")
+    mch_resource.wait_for_phase("Running")
+    backup_pod_status_check()
+
+
+def create_s3_bucket(access_key, secret_key, bucket_name):
```
Comment (Shrivaibavi) [inline ocs_ci/helpers/dr_helpers.py:1129]: In automation do we use s3 by default and not noobaa ?

File: ocs_ci/helpers/dr_helpers.py [hunk 12]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    run_cmd(f"oc create -f {oadp_yaml.name}")
+    # Validation
+    validate_dpa()
+
+
+@retry((CommandFailed, ACMClusterConfigurationException), tries=10, delay=10)
+def validate_dpa():
+    """
+    Validate
+    1. 3 restic / Node pods
```
Comment (Shrivaibavi) [inline ocs_ci/helpers/dr_helpers.py:1269]: 1. 3 restic / Node-agent pods

File: ocs_ci/helpers/dr_helpers.py [hunk 13]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+
+    if version.compare_versions(f"{oadp_version} >= 1.2"):
+        restic_or_node_pod_prefix = "node-agent"
+    else:
+        restic_or_node_pod_prefix = "restic"
+    restic_or_node_list = get_pods_having_label(
+        f"name={restic_or_node_pod_prefix}", constants.ACM_HUB_BACKUP_NAMESPACE
+    )
+    if len(restic_or_node_list) != constants.RESTIC_OR_NODE_POD_COUNT:
+        raise ACMClusterConfigurationException("restic/node pod count mismatch")
```
Comment (Shrivaibavi) [inline ocs_ci/helpers/dr_helpers.py:1285]: Please rename as node-agent pod instead of node pod

File: ocs_ci/helpers/dr_helpers.py [hunk 14]
```diff
@@ -1083,3 +1098,258 @@ def verify_drpolicy_cli(switch_ctx=None):
...
+    for pod in restic_or_node_list:
+        if pod["status"]["phase"] != "Running":
+            raise ACMClusterConfigurationException(
+                "restic/node pod not in 'Running' phase"
```
Comment (Shrivaibavi) [inline ocs_ci/helpers/dr_helpers.py:1289]: same as above

File: ocs_ci/ocs/constants.py [hunk 15]
```diff
@@ -933,15 +933,16 @@
 RDR_MODE = "regional-dr"
 MDR_MODE = "metro-dr"
 MDR_DR_POLICY = "odr-policy-mdr"
-MDR_RESTIC_POD_COUNT = 3
-MDR_VELERO_POD_COUNT = 1
+RESTIC_OR_NODE_POD_COUNT = 3
```
Comment (Shrivaibavi) [inline ocs_ci/ocs/constants.py:936]: same as above

File: ocs_ci/helpers/dr_helpers.py (hunk 6 above)
Comment (Shrivaibavi) [reply ocs_ci/helpers/dr_helpers.py:1312]: This check can be performed post backupschedule creation and not when DPA validation i guess, Please correct me if i am wrong

File: ocs_ci/helpers/dr_helpers.py (hunk 11 above)
Comment (prsurve) [reply ocs_ci/helpers/dr_helpers.py:1129]: Yes we use s3 coming from aws

**Review by am-agrawa — DISMISSED** [review]

File: ocs_ci/deployment/deployment.py [hunk 16]
```diff
@@ -12,10 +12,11 @@
 import time
 from pathlib import Path
 import base64
+
```
Comment (am-agrawa) [inline ocs_ci/deployment/deployment.py:15]: extra line

**Review by am-agrawa — APPROVED** [review]

**Review by keesturam — APPROVED** [review]

**Pull request merged**